py/picca/cf.py

```diff
@@ -596,6 +686,201 @@ def compute_dmat_forest_pairs(log_lambda1, log_lambda2, r_comov1, r_comov2,
+                # first eta, third term: (non-zero only for order=1)
+                #   weight*(Lambda-bar(Lambda))*(Lambda-bar(Lambda))/
+                #   sum(weight*(Lambda-bar(Lambda)**2))
+                # second eta, first term: kronecker delta
+                eta4[j+num_pixels2*model_bins] += (weights1[i]*log_lambda_minus_mean1[i] /
+                        sum_weights_square_log_lambda_minus_mean1)
+                # first eta, third term: (non-zero only for order=1)
+                #   weight*(Lambda-bar(Lambda))*(Lambda-bar(Lambda))/
+                #   sum(weight*(Lambda-bar(Lambda)**2))
+                # second eta, second term: weight/sum(weights)
+                eta7[model_bins] += (weights2[j]/sum_weights2 * 
+                                     (weights1[i]*log_lambda_minus_mean1[i] /
+                                     sum_weights_square_log_lambda_minus_mean1))
+                if order2 == 1:
+                    # first eta, third term: (non-zero only for order=1)
+                    #   weight*(Lambda-bar(Lambda))*(Lambda-bar(Lambda))/
+                    #   sum(weight*(Lambda-bar(Lambda)**2))
+                    # second eta, third term: (non-zero only for order=1)
+                    #   weight*(Lambda-bar(Lambda))*(Lambda-bar(Lambda))/
+                    #   sum(weight*(Lambda-bar(Lambda)**2)
+                    eta8[model_bins] += ( weights1[i]*log_lambda_minus_mean1[i] * 
+                                          weights2[j]*log_lambda_minus_mean2[j] /
+                                          sum_weights_square_log_lambda_minus_mean1 /
+                                          sum_weights_square_log_lambda_minus_mean2 )
+
+    # Now add all the contributions together
+    unique_bins_model = np.unique(all_bins_model)
+    for pair in range(num_pairs):
+        i = all_i[pair]
+        j = all_j[pair]
+        bins = all_bins[pair]
+        model_bins = all_bins_model[pair]
+        weights12 = weights1[i] * weights2[j]
+        # first eta, first term: kronecker delta
+        # second eta, first term: kronecker delta
+        dmat_bin = model_bins + num_model_bins_r_par * num_model_bins_r_trans * bins
+        dmat[dmat_bin] += weights12
+        # rest of the terms
+        for k in unique_bins_model:
+            dmat_bin = k + num_model_bins_r_par * num_model_bins_r_trans * bins
+            dmat[dmat_bin] += (
+                weights12 *
+                (eta5[k] + 
+                + eta6[k] * log_lambda_minus_mean2[j] + 
+                + eta7[k] * log_lambda_minus_mean1[i] + 
+                + eta8[k] * log_lambda_minus_mean1[i] * log_lambda_minus_mean2[j] 
+                - eta1[i+num_pixels1*k] 
+                - eta2[j+num_pixels2*k] 
+                - eta3[i+num_pixels1*k] * log_lambda_minus_mean2[j] 
+                - eta4[j+num_pixels2*k] * log_lambda_minus_mean1[i])
+                )
 
 def compute_metal_dmat(healpixs, abs_igm1="LYA", abs_igm2="SiIII(1207)"):
```

Comment:
I take it we don't have the fast version for this function, do we? Is it planned for a later commit?

Reply:
There were no warnings on this one. I don't plan to update it for now. But you can add this to the todo list for other people. It's not easy though.